### DISCUSSION TITLE
ESP32 with ADC_VCC setting - how to configure in Web UI?

### DISCUSSION THREAD

**Discussion by belveder79** [arendst/Tasmota#21683] [Q&A] [answered]
This is probably a stupid question, but how do I enable the VCC reading on an ESP32 over an analog pin with the internal 
`#define USE_ADC_VCC` switch? For me the Vcc simply won't show up.

1. From the [Documentation](https://tasmota.github.io/docs/ADC/#adc_vcc) it seems that there is no additional wiring required, right? 
2. So the actual supply voltage is read internally and provided, correct?
3. I have gpios 34,35,36 and 39 as analog input available. What is the correct setting on the WebUI? Is setting it required at all?

There are the usual options like Analog Input, MQ etc., but no extra option. Whatever I set for example for 34, it gives me some readings that don't make sense. 

Can someone please clarify? Thx

**Comment by sfromis** [comment, marked as answer]
You cannot make any web UI config for that. It is a feature on ESP8266 (not ESP32), where you can do a build with `#define USE_ADC_VCC` to enable it, basically instead of normal analog.

What you can do on ESP32 is to wire an analog input pin to the 3.3V power input. If what you want is *instead* battery voltage, you can wire it to the battery output, where you need to use a voltage divider, as ESP32 analog inputs are only up to 3.3V. In both cases, you can config the analog input for range, and set corresponding `AdcParam` values for a linear fitting, if you need to adjust the raw numbers.

**Reply by belveder79** [reply]
Ah, ok, got it. 
So it is only for the ESP8266 with the #define, and on the ESP32 needs an extra wire to a respective pin and a proper config.
Thx for clarification...

**Comment by belveder79** [comment]
I need to reopen this one, because it does not work for me for whatever reason on stock 14.1.0....

I tried to hook up a voltage divider to the LiIon Battery shield with 220k/60k which should give me some 3.3V or so max on the ADC pin, which I expected to scale to somewhat around 4.2V at 4095 and 2.8V at 2750 or so. The problem is, the ADC Input does not change AT ALL. I added a half-empty LiIon cell at about 4V and it gave me 3.915, and that's it. no matter the charge is, it stays like this. I have a second assembly with some 470k/130k resistors, same game, somewhere stuck at 3700 something with 4.0V, no change. I know that the AD converter on the ESP32 is not the best but is it really so bad?

**Reply by sfromis** [reply]
I'd verify the setup with first connecting the basic 3.3V directly to the analog pin (without anything else), just to make sure that the basics are right. When I do that (3.28V), I get 3908, which is "conspicuously close" to the value you report, suggesting that the wiring may be off somewhere.

You could also connect the analog output with Berry code (or a pot, which is basically another voltage divider), to vary values, again for "sanity checking".

**Reply by belveder79** [reply]
yes, what I will do is provide different values with a variable power supply, see if the values change if I also change the voltage. Will report asap...

**Reply by belveder79** [reply]
so it seems to me that it just maxes and mins out at a specific value, which is different for each ESP32 and voltage level. I supplied 3.20 volts sharp and it gives me 3882 as a value. Anything above does not change the value any more. Same with bottom - below 0.02 it does not change and gets never lower than 176.

```
3.30 3882
3.20 3882
3.07 3784
2.53 3118
1.84 2284
1.40 1736
0.63 773
0.21 257
0.02 176
0.0 176
```

So it seems that the plausible range to do any measurements is 0.1V to 3.2V or so, will adapt my divider accordingly...

**Reply by sfromis** [reply]
I'd suggest avoiding a setup where you need to get close to either minimum or maximum; the ADC is known to be "less accurate" at the ends.

**Reply by belveder79** [reply]
yes, tuned it now with 87k/220k to go from somewhere 3.0V downwards... seems to work, thx....

**Comment by ChrisVonFrosta** [comment]
hi. i came around with the same behavior. ADC Input has 176, even if i connect the pin directly to GND.
my windspeed sensor has 0-5V for 0-30 m/s. So counting from 0,0V is more important than being accurate at the upper limit.
is ther no solution to come lower than 176 on ADC?
